Parse desc/name and desc/id responses into Stem's RelayDescriptor class

sammyshj opened the pull request:
Implemented _get_descriptor() method
The parameters for the method are, `nickname`, `fingerprint`. These are the relay's nickname/fingerprint for which we want the descriptor information.
Returns the descriptor information. If use_stem is set to True, a `stem.descriptor.server_descriptor.RelayDescriptor` object is returned with the descriptor information.

sammyshj commented:
@meejah Let me know if the method looks fine. If everything looks good then, I will go ahead and add tests for the method.

meejah commented:
To work with the rest of the magic accessors of TorInfo, you'll need to arrange for this method to be called when accessing ``desc.name.<descriptor name>`` or ``desc.id.<descriptor fingerprint>`` -- once TorInfo is setup it hides all members except via the dotted-names things.

So -- you should be able to access descriptors with TorInfo right now on master without any changes...you'll just want to intercept those accesses in case ``use_stem`` is True in the underlying protocol and return Stem objects instead.

sammyshj commented:
@meejah I do not understand the MagicContainer class completely and hence I am having trouble converting the `_get_descriptor()` method into something like, `desc.name.<descriptor name>` as you mentioned in the above comment.
Can you shed more light on the conversion to dotted object please?

meejah commented:
Well, the above dotted-stuff should work right now I believe; basically the magic-container is translating those into "GETINFO desc/name/<arg to the method>" if the ``info/names`` ended with a ``*`` or simply into "GETINFO foo/bar" (from ``foo.bar``) if the meta-information about that key did *not* end with a *.

So essentially during setup you'll have to figure out if use_stem is set, and if so intercept the setup of certain keys and replace with a suitable other method (so you'd be intercepting "desc/name/*" and "desc/id/*" or so).

sammyshj commented:
@meejah I have added a new commit, https://github.com/meejah/txtorcon/commit/de88b1614ec842e3704924c4254d99aa47bf40b8
In this commit, I have added a new _get_descriptor method which is called when the use_stem flag is set to `True` and the `desc.name()` or `desc.id()` is called.
Let me know if this is what is required, then, I will add appropriate comments for the method, tests for the method and squash all the commits into one.

meejah commented on txtorcon/torinfo.py:
```diff
@@ -66,6 +66,14 @@ def dump(self, prefix):
+        pass
+    return data
+
 class ConfigMethod(object):
```
This all looks pretty good (haven't actually tried it yet, though)!

I think probably the ImportError should just be allowed to happen, as you only add ``_get_descriptor`` if .use_stem was True -- and then it's an error if you don't have Stem installed, I'd say? Also maybe "_parse_stem_descriptor" or something instead for a helper name, since we're not really getting it here (just parsing).

Thanks!

sammyshj replied:
Yes, the ImportError will be shown then, when the .use_stem is True and we don't have Stem installed. Also, I am adding the comment for the method.

sammyshj commented:
@meejah Added test for the _parse_stem_descriptor method. If the code looks good, I will squash all the commits into one.
Thanks

meejah commented on txtorcon/torinfo.py:
```diff
@@ -99,6 +109,9 @@ def stripper(key, arg):
+        descriptor_keys = ["desc/name", "desc/id"]
+        if self.info_key in descriptor_keys and self.proto.use_stem:
+            d.addCallback(_parse_stem_descriptor)
         return d
```
Could we just make the call back the RelayDescriptor ctor itself?

sammyshj replied:
Does this mean that we do not have a _parse_stem_descriptor method and, just directly return the RelayDescriptor object as a callback?

meejah replied:
I meant that we can probably just use the "RelayDescriptor" constructor as the callback, directly (since all we do in _parse_stem_descriptor is call RelayDesciptor(data)). Might actually be better to just leave the method you have, though, in case it needs more stuff in future ...?

sammyshj replied:
I think its better to have the RelayDescriptor constructor as the callback. This is because all the function is doing is just calling RelayDescriptor(data). The changes will be seen in the next commit.

sammyshj replied:
Updated in the commit https://github.com/sammyshj/txtorcon/commit/dc6650c46e37864575b7f7e305dfdc97a2d5a1ea
Should I squash all the commits into one now?

meejah commented on test/test_torinfo.py:
```diff
@@ -118,6 +121,46 @@ def test_with_arg(self):
+        descriptor_info = yield d
+        self.assertEqual(type(descriptor_info).__name__, 'RelayDescriptor')
+
 
```
Should usually use isinstance instead of comparing types

meejah commented:
yes, looks great! 

...and yes please squash (and just change the commit-comment to a description of the whole change instead of leaving the individual commit-comments).

sammyshj commented:
@meejah Made the required changes. This PR can be merged now. Thanks.

meejah commented:
looks great, thanks for the patch :)
